### PR TITLE
Catch retrieveToken errors

### PR DESCRIPTION
**Pull request by Lehcar**
## Description

This adds a try-catch to `await core.getIDToken(githubAudience)` and to the part of `generateJwt(privateKey, keyPassword, Number(tokenTtl));` that makes an external service call. 

We actually just use `core.getIDToken`, but I figured I might as well as another try-catch.

`core.getIDToken` actually does up to 10 retries automatically ([here](https://github.com/actions/toolkit/blob/63c648f3c20feb23a4251afd9b3574570d38748a/packages/core/src/oidc-utils.ts#L11-L19)), so we don't need to add retries here. But I'm hoping a try-catch here will reduce the `triggerUncaughtException(err, true /* fromPromise */)` errors and help us debug.

I also added a test for authenticating using `jwt` and GitHub OIDC.

## Testing
Using https://github.com/github/internal-actions/pull/332 to test

cc\ https://github.com/github/build-systems/issues/3307

### CODE REVIEW

File: src/auth.js [hunk 1]
```diff
@@ -34,7 +34,14 @@ async function retrieveToken(method, client) {
             const githubAudience = core.getInput('jwtGithubAudience', { required: false });
 
             if (!privateKey) {
-                jwt = await core.getIDToken(githubAudience)
+                try {
+                    core.debug("Fetching ID token from GitHub OIDC provider")
+                    jwt = await core.getIDToken(githubAudience)
+                } catch (err) {
+                    core.error("Error fetching ID token from GitHub OIDC provider")
```
Comment (Lehcar) [inline src/auth.js]: I'm not sure if it'd be better to be more explicit here with the messaging. For example, should I change this to "Error calling core.getIDToken"?

**Comment by Lehcar** [comment]
I'm setting the target branch of this PR to [allan-pires/throw-error-at-exportsecrets-step](https://github.com/wagnerm/vault-action/tree/allan-pires/throw-error-at-exportsecrets-step) for the diff (since that's what we're currently using), but I think this should be merged into `main` once we're finished reviewing. 
Thoughts?

File: src/auth.js (hunk 1 above)
Comment (wagnerm) [reply src/auth.js]: I guess I'm wondering what the goal of catching this error is? Does it happen currently or is being hidden? If we aren't doing something with the error and just re-throwing it, will it be hidden again?

Comment (Lehcar) [reply src/auth.js]: I'm not sure if it's currently happening or being hidden. I've seen some `triggerUncaughtException(err, true /* fromPromise */)` errors in the logs and that's caused by not catching errors thrown by promises.
I believe we have to rethrow it in order for the workflow to fail.

I'm basically copying [this change](https://github.com/hashicorp/vault-action/commit/8708b58476ed098be5f1169c405233c2850ec893) Allan added to get more error output, but I'm going a level down.

This is also adding an error annotation, so I'm not sure if we want that or if we want `core.setOutput`.

Comment (wagnerm) [reply src/auth.js]: >  errors in the logs and that's caused by not catching errors thrown by promises.
I believe we have to rethrow it in order for the workflow to fail.

Is there any way we test this manually or in the code to make sure? I'm really not familiar with JS at all so I don't have a feel for what is really happening. If you know this will uncover those errors then that is fine too.

> This is also adding an error annotation, so I'm not sure if we want that or if we want core.setOutput

I think the reason to use `core.setOutput("errormessage", error?.message);` is that we can then inspect the errormessage output outside of the action step. For example https://github.com/github/internal-actions/blob/main/.github/workflows/default.yml#L247

In this instance, I'm not sure `core.setOutput(err?.message)` is correct because it is missing a variable to bind the output to, so I think we might as well make that `core.setOutput("errormessage", err?.message);`?

Comment (wagnerm) [reply src/auth.js]: Also wanted to say, we might want to adjust the post processing code to capture and send metrics for the error you are generating, otherwise we will not be able to easily see if these errors are occurring unless we do some manual inspection.

Comment (Lehcar) [reply src/auth.js]: > Is there any way we test this manually or in the code to make sure? I'm really not familiar with JS at all so I don't have a feel for what is really happening. If you know this will uncover those errors then that is fine too.
>
I'm trying to find node docs on this, but I got this conclusion from https://stackoverflow.com/questions/71077701/react-jest-nodeinternal-process-promises246-triggeruncaughtexceptionerr-tr. `core.getIDToken` can throw errors so I think we should be catching these anyways. https://github.com/actions/toolkit/blob/c5c786523e095ca3fabfc4d345e16242da34e108/packages/core/src/oidc-utils.ts

Maybe, in this case, we should catch this error, and append our own error message before we rethrow (instead of just throwing the original message)?

> I think the reason to use core.setOutput("errormessage", error?.message); is that we can then inspect the errormessage output outside of the action step. For example https://github.com/github/internal-actions/blob/main/.github/workflows/default.yml#L247
>
> In this instance, I'm not sure core.setOutput(err?.message) is correct because it is missing a variable to bind the output to, so I think we might as well make that core.setOutput("errormessage", err?.message);?
>
I think I was mixing up `core.setOutput` and `console.log` 😆 My js is rusty

Comment (Lehcar) [reply src/auth.js]: Yeah, since [`entry.js` will set the `errormessage` output](https://github.com/wagnerm/vault-action/blob/allan-pires/throw-error-at-exportsecrets-step/src/entry.js), I think it might be better to throw the new error (and have it contain the old error). I will make that change 👍 

I'm going to manually test this by either manually throwing an error or by setting one of the inputs to gibberish

Comment (Lehcar) [reply src/auth.js]: I tested manually, and I think wrapping the error message is better. I'm still not sure where the `triggerUncaughtException(err, true /* fromPromise */)` came from, maybe from somewhere in `getIDToken`? (I forgot to copy the link when I saw that initial error 😭 )
### Test throwing error with the error handling
https://github.com/github/internal-actions/actions/runs/6165865722/job/16734398929
image

### Test throwing error without the error handling
https://github.com/github/internal-actions/actions/runs/6165895752/job/16734479869?pr=332
image

Comment (wagnerm) [reply src/auth.js]: After looking at the testing, with the error handling it kind of looks like we are hiding the original location of the failure? Do we want to do that?

Comment (Lehcar) [reply src/auth.js]: I'm not sure if the original error location helps us out that much here, especially in the case of a `triggerUncaughtException(err, true /* fromPromise */` error.

Comment (Lehcar) [reply src/auth.js]: Honestly, I can't tell if the original error location is getting hidden because my test error was a couple lines above it and I had a bunch of debugs 🤣

Comment (Lehcar) [reply src/auth.js]: I'm not sure though if this PR addresses that issue. The original error originated in node, and I don't have the link (I forgot to link it 😢 ). Should we just close this?

Comment (Lehcar) [reply src/auth.js]: I'm updating this to just log and rethrow the original error

**Comment by Lehcar** [comment]
spoke with Max offline. We shouldn't merge this PR with main because it makes it harder to sync from the source repo

**Comment by Lehcar** [comment]
I'm going to close this. The integration tests won't work without updating our current vault-action with the original hashicorp action, which is something we want to avoid at this time. 

Maybe we can add this later when this gets updated

**Comment by Lehcar** [comment]
I found an example of the `triggerUncaughtException(err, true /* fromPromise */);` happening today ([link](https://github.com/github/github/actions/runs/6315216227/job/17147312468)): 

```
Error: Response code 403 (Forbidden)
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

HTTPError: Response code 403 (Forbidden)
    at Request.<anonymous> (/home/root/work/_actions/wagnerm/vault-action/8708b58476ed098be5f1169c405233c2850ec893/dist/index.js:3702:42)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  code: 'ERR_NON_2XX_3XX_RESPONSE',
  timings: {
    start: 169574[36](https://github.com/github/github/actions/runs/6315216227/job/17147312468#step:12:38)95649,
    socket: 16957[43](https://github.com/github/github/actions/runs/6315216227/job/17147312468#step:12:45)695649,
    lookup: 1695743695651,
    connect: 1695743695653,
    secureConnect: 1695743695656,
    upload: 1695743695656,
    response: 1695743695673,
    end: 1695743695674,
    error: undefined,
    abort: undefined,
    phases: {
      wait: 0,
      dns: 2,
      tcp: 2,
      tls: 3,
      request: 0,
      firstByte: 17,
      download: 1,
      total: 25
    }
  }
}
```
I checked [that line](https://github.com/wagnerm/vault-action/blob/8708b58476ed098be5f1169c405233c2850ec893/dist/index.js#L3702) in `dists/index.js` and that just looks like the general promise handling code to me.

I do think we need to add these try-catch blocks around our awaits so we know which async calls failed, so I'm reopening this. I'm going to see if I can just append the old error message and stack trace to a new error message.


cc\ https://github.com/github/build-systems/issues/3307#issuecomment-1736418110

File: src/auth.js [hunk 2]
```diff
@@ -34,7 +35,14 @@ async function retrieveToken(method, client) {
...
 
             if (!privateKey) {
-                jwt = await core.getIDToken(githubAudience)
+                try {
+                    core.debug("Fetching ID token from GitHub OIDC provider")
+                    jwt = await core.getIDToken(githubAudience)
+                } catch (err) {
+                    const new_error = new Error(`Error fetching ID token from GitHub OIDC provider., message: ${err.message}`);
+                    new_error.stack += err.stack;
+                    throw new_error;
```
Comment (wagnerm) [inline src/auth.js:44]: I accidentally deleted my comment from above because I meant to put it here. It's probably in your email. But I was advocating for just logging the error and rethrowing instead of stacking errors.

File: src/auth.js [hunk 3]
```diff
@@ -88,8 +96,14 @@ function generateJwt(privateKey, keyPassword, ttl) {
...
-    const decryptedKey = rsasign.KEYUTIL.getKey(privateKey, keyPassword);
-    return rsasign.KJUR.jws.JWS.sign(alg, JSON.stringify(header), JSON.stringify(payload), decryptedKey);
+    try{
+        const decryptedKey = rsasign.KEYUTIL.getKey(privateKey, keyPassword);
+        return rsasign.KJUR.jws.JWS.sign(alg, JSON.stringify(header), JSON.stringify(payload), decryptedKey);
+    } catch (err) {
+        const new_error = new Error(`Error generating the jwt., message: ${err?.message}`);
+        new_error.stack += err.stack;
+        throw new_error;
+    }
```
Comment (wagnerm) [inline src/auth.js]: Do we know this fails? This is a local operation afaik that doesn't await so I'm not sure how it relates to the promises problem

Comment (Lehcar) [reply src/auth.js]: I'll remove it. We don't call it, and I don't remember why I added it

File: src/auth.js (hunk 2 above)
Comment (Lehcar) [reply src/auth.js:44]: I think it makes more sense to stack the errors since we're catching all these errors and processing them further up in the stack.

Comment (Lehcar) [reply src/auth.js:44]: Also, we'll get a full stack trace with this method.

File: dist/index.js [hunk 4]
```diff
@@ -18768,7 +18769,14 @@ async function retrieveToken(method, client) {
             const githubAudience = core.getInput('jwtGithubAudience', { required: false });
 
             if (!privateKey) {
-                jwt = await core.getIDToken(githubAudience)
+                try {
+                    core.debug("Fetching ID token from GitHub OIDC provider")
+                    jwt = await core.getIDToken(githubAudience)
+                } catch (err) {
+                    const new_error = new Error(`Error fetching ID token from GitHub OIDC provider., message: ${err.message}`);
```
Comment (wagnerm) [inline dist/index.js:18776]: Is there a reason this is a const?

Comment (Lehcar) [reply dist/index.js:18776]: Leftover habit from coding in typescript at my last job. We used to have a linter that would yell at us if we didn't use const 😆 
const and let are generally best practices for javascript over var ever since they were introduced in es6: https://www.freecodecamp.org/news/differences-between-var-let-const-javascript/
https://medium.com/javascript-scene/javascript-es6-var-let-or-const-ba58b8dcde75

Basically, since I'm using `const` here, another error object can't get assigned to the `new_error` variable. You can still update properties of the object, but the original object pointer can't change. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const

**Review by wagnerm — APPROVED** [review]
worth a try

**Comment by Lehcar** [comment]
Spoke to Max about how/where to merge this in Slack. We're not going to merge it into the main/default branch at the moment because it'll make it harder for us to synchronize this fork with the original hashicorp repo. Instead I'm just going to merge this into Allan's branch.
